modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java

```diff
@@ -6347,4 +6491,137 @@ protected String lockRetryErrorMessage(String op) {
             getClass().getSimpleName() + " [grpName=" + grpName + ", treeName=" + name() + ", metaPageId=" +
             U.hexLong(metaPageId) + "].";
     }
+
+    /**
+     * The operation of deleting a range of values.
+     * <p>
+     * Performs the removal of several elements from the leaf at once.
+     */
+    protected class RemoveRange extends Remove {
+        /** Upper bound. */
+        private final L upper;
+
+        /** Lower bound. */
+        private final L lower;
+
+        /** List of removed rows. */
+        private final List<L> removedRows;
+
+        /** The number of remaining rows to remove ({@code -1}, if the limit hasn't been specified). */
+        private int remaining;
+
+        /** Flag indicating that no more rows were found from the specified range. */
+        private boolean completed;
+
+        /** The index of the highest row found on the page from the specified range. */
+        private int highIdx;
+
+        /**
+         * @param lower Lower bound (inclusive).
+         * @param upper Upper bound (inclusive).
+         * @param needOld {@code True} If need return old value.
+         */
+        protected RemoveRange(L lower, L upper, boolean needOld, int limit) {
+            super(lower, needOld, rmvRangeFromLeaf);
+
+            this.lower = lower;
+            this.upper = upper;
+
+            remaining = limit <= 0 ? -1 : limit;
+            removedRows = needOld ? new LinkedList<>() : null;
```

Comment:
Not sure, but perhaps ArrayList will be more effective here (less GC pressure)

Reply:
Done